src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -299,9 +300,10 @@
                      (:id instance)
                      max-requests)
           (flush-instance-fn pool instance))
-        (.putFirst pool instance)))
-    ;; if we get here, we got a Retry, so we just put it back into the pool.
-    (.putFirst (:pool instance) instance)))
+        (.returnItem pool instance)))
+    ;; if we get here, it was from a borrow and we got a Retry, so we just
+    ;; return it to the pool.
+    (.returnItem (:pool instance) instance)))
```

Comment:
I think it's probably functionally equivalent, but, technically I think we're putting a Pill back into the pool here, so I wonder if just in terms of semantics it would be better to call your {{insertPill}} method?